File: 2025/05/cpp/05.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long      ll;
typedef vector<int>    vi;
typedef vector<bool>   vb;
typedef pair<int, int> ii;

const int INF = 0x3f3f3f3f;
const ll  INFLL = 0x3f3f3f3f3f3f3f3fll;

#define each(x, s)  for (auto& x : s)
#define loop(x)     for (int i = 0;i < x;i++)
#define vloop(v, x) for (int v = 0;v < x;v++)
#define iter(a)     a.begin(), a.end()
#define riter(a)    a.rbegin(), a.rend()
#define endl        "\n"

int task1(vector<pair<ll, ll>> rs, vector<ll> is) {
  vector<pair<ll, ll>> sl;
  for (auto& [l, r] : rs) {
    sl.emplace_back(l, -1);
    sl.emplace_back(r, 1);
  }
  each(x, is) sl.emplace_back(x, 0);
  int cnt = 0;
  int ans = 0;
  sort(iter(sl));
  for (auto& [x, t] : sl) {
    if (!t) ans += !!cnt;
    else cnt -= t;
  }
  return ans;
}

ll task2(vector<pair<ll, ll>> rs) {
  sort(iter(rs));
  ll last = -INF;
  ll ans = 0;
  for (auto& [l, r] : rs) {
    if (r <= last) continue;
    if (l <= last + 1) {
      ans += r - last;
    }
    else {
      ans += r - l + 1;
    }
    last = r;
  }
  return ans;
}

int main() {
  ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);
  
  string s;
  vector<pair<ll, ll>> rs;
  vector<ll> is;
  while (getline(cin, s) and !s.empty()) {
    stringstream ss;
    ss << s;
    ll l, r;
    char c;
    ss >> l >> c >> r;
    rs.emplace_back(l, r);
  }
  ll x;
  while (cin >> x) is.push_back(x);
  cout << "Task 1: " << task1(rs, is) << endl;
  cout << "Task 2: " << task2(rs) << endl;
  return 0;
}

File: 2025/05/rust/05.rs
use std::io;

fn task1(rs: &Vec<(i64, i64)>, is: &Vec<i64>) -> i64 {
  let mut ri = 0;
  let mut ii = 0;
  let mut ans = 0;
  while ri < rs.len() && ii < is.len() {
    let (l, r) = rs[ri];
    let x = is[ii];
    if x < l {
      ii += 1;
    }
    else if x > r {
      ri += 1;
    }
    else {
      ans += 1;
      ii += 1;
    }
  }
  ans
}

fn task2(rs: &Vec<(i64, i64)>) -> i64 {
  let mut ans = 0;
  let mut last = -1;
  for (l, r) in rs {
    if *l > last + 1 {
      ans += r - l + 1;
      last = *r;
    }
    else if *r > last {
      ans += r - last;
      last = *r;
    }
  }
  ans
}

fn main() {
  let mut rs = Vec::new();
  let mut is = Vec::new();
  for line in io::stdin().lines() {
    let line = line.unwrap();
    if line.is_empty() {
      break;
    }
    let (l, r) = line.split_once('-').unwrap();
    rs.push((l.parse::<i64>().unwrap(), r.parse::<i64>().unwrap()));
  }
  for line in io::stdin().lines() {
    let line = line.unwrap();
    is.push(line.parse::<i64>().unwrap());
  }
  rs.sort();
  is.sort();
  println!("Task 1: {}", task1(&rs, &is));
  println!("Task 2: {}", task2(&rs));
}
